generate.py

```diff
@@ -0,0 +1,208 @@
+# Copyright 2022 The UK Pilot Synthetic Data Gen Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Script to generate impressions for a single Campaign."""
+
+
+from dataclasses import dataclass, asdict
+import datetime
+import random
+import numpy as np
+from scipy import stats
+import operator
+import inspect
+import csv
+import math
+from functools import reduce
+
+
+# Total number of Virtual people, per the UK population
+NUM_VIDS = 60_000_000
+
+# Daily uniform noise size as fraction of the impressions that day.
+DAILY_NOISE_FACTOR = 0.1
+
+# Acceptable change applied to max frequency to make the num impressions
+# play well with freq distribution.
+ACCEPTABLE_FREQ_DIST_CORRECTION_FACTOR = 0.1
+
+
+class CampaignSpec:
+    """Class that samples impressions given specification for a single campaign.
+
```

Comment:
Before going to details, you may summarzie the functionality of this class as "Samples impressions on a given EDP on given dates, such that they approximately align with the given number of impressions, reach, and distributions of frequency, video completion, and viewability."

Reply:
awesome, thank you